lib/new_relic/agent/instrumentation/sinatra.rb

```diff
@@ -23,165 +27,57 @@
   end
 
   executes do
-    ::Sinatra::Base.class_eval do
-      include NewRelic::Agent::Instrumentation::Sinatra
-
-      alias dispatch_without_newrelic dispatch!
-      alias dispatch! dispatch_with_newrelic
+    if use_prepend?
+      ::Sinatra::Base.class_eval do
+        class << self
+          include ::NewRelic::Agent::Instrumentation::ControllerInstrumentation
+          include NewRelic::Agent::Instrumentation::Sinatra
+        end
+        include NewRelic::Agent::Instrumentation::Sinatra
 
-      alias process_route_without_newrelic process_route
-      alias process_route process_route_with_newrelic
+        register NewRelic::Agent::Instrumentation::Sinatra::Ignorer
+        # chain_instrument NewRelic::Agent::Instrumentation::SinatraInstrumentation::Chain
+      end
+      ::Sinatra.module_eval do
+        register NewRelic::Agent::Instrumentation::Sinatra::Ignorer
+      end
 
-      alias route_eval_without_newrelic route_eval
-      alias route_eval route_eval_with_newrelic
+      prepend_instrument ::Sinatra::Base, NewRelic::Agent::Instrumentation::SinatraInstrumentation::Prepend
+    else
 
-      register NewRelic::Agent::Instrumentation::Sinatra::Ignorer
+      chain_instrument NewRelic::Agent::Instrumentation::SinatraInstrumentation::Chain
     end
 
-    module ::Sinatra
-      register NewRelic::Agent::Instrumentation::Sinatra::Ignorer
-    end
   end
 
-  executes do
+  # had to keep this chunk in an executes block bc rack
+  executes do 
     if Sinatra::Base.respond_to?(:build)
       # These requires are inside an executes block because they require rack, and
       # we can't be sure that rack is available when this file is first required.
       require 'new_relic/rack/agent_hooks'
       require 'new_relic/rack/browser_monitoring'
-
-      ::Sinatra::Base.class_eval do
-        class << self
-          alias build_without_newrelic build
-          alias build build_with_newrelic
+      if use_prepend?
+        prepend_instrument ::Sinatra::Base.singleton_class, NewRelic::Agent::Instrumentation::SinatraInstrumentation::PrependBuild
+      else
+        ::Sinatra::Base.class_eval do
```

Comment:
move this to approriately named Chain module with an #instrument! method so implementation details are removed fromthe Dependency Detection file.